### ISSUE TITLE
Talon occasionally stops parsing any utterances

### ISSUE THREAD

**Issue by tuomassalo** [talonvoice/talon#4]
I have encountered this problem a few times now. After running Talon for some time (hours) and making (eventually valid) changes to .py files, Talon stops parsing anything. There's nothing else in talon.log or draconity output.

There might be two different things. Case one:
- my debug.py still prints "IO phrase []" for everything I say, but I get no words

~~Case two:~~
~~- debug.py prints nothing~~
~~- Talon starts using lots of CPU (e.g. 80 %)~~ - this was caused by an obscure infinite loop in my .py file.

I'm running High Sierra, Dragon 6.0.8, latest Talon.

How could I provide more debugging information? What to do when I encounter the problem next time?

**Comment by tuomassalo** [comment]
Still happens on 0.0.6.0. Tailing `~/.talon/talon.log`:

```2018-03-05 15:07:59    IO [grammar reloaded]
2018-03-05 15:08:04 ERROR cb error <class 'talon.ui.Mon'> topic="app_activate" cb=<bound method ContextGroup.ui_event of <talon.voice.ContextGroup object at 0x000000011773f6e0>>
Traceback (most recent call last):
  File "/Applications/Talon.app/Contents/Resources/pypy/site-packages/talon/dispatch.py", line 58, in safe_call
    return self.call(cb, args)
  File "/Applications/Talon.app/Contents/Resources/pypy/site-packages/talon/dispatch.py", line 54, in call
    return cb(*args)
  File "/Applications/Talon.app/Contents/Resources/pypy/site-packages/talon/voice.py", line 251, in ui_event
    self.update_scope()
  File "/Applications/Talon.app/Contents/Resources/pypy/site-packages/talon/voice.py", line 259, in update_scope
    self.reload()
  File "/Applications/Talon.app/Contents/Resources/pypy/site-packages/talon/voice.py", line 245, in reload
    self.grammar.listset(name, items)
  File "/Applications/Talon.app/Contents/Resources/pypy/site-packages/talon/grammar.py", line 100, in listset
    resp = self.cmd('g.listset', list=name, items=words)
  File "/Applications/Talon.app/Contents/Resources/pypy/site-packages/talon/grammar.py", line 113, in cmd
    return self.engine.cmd(cmd, **kwargs)
  File "/Applications/Talon.app/Contents/Resources/pypy/site-packages/talon/engine.py", line 82, in cmd
    raise EngineCmdErr(f'engine command "{cmd}" failed')
talon.engine.EngineCmdErr: engine command "g.listset" failed
2018-03-05 15:08:14 ERROR cb error <class 'talon.ui.Mon'> topic="app_activate" cb=<bound method ContextGroup.ui_event of <talon.voice.ContextGroup object at 0x000000011773f6e0>>
Traceback (most recent call last):
  File "/Applications/Talon.app/Contents/Resources/pypy/site-packages/talon/dispatch.py", line 58, in safe_call
    return self.call(cb, args)
  File "/Applications/Talon.app/Contents/Resources/pypy/site-packages/talon/dispatch.py", line 54, in call
    return cb(*args)
  File "/Applications/Talon.app/Contents/Resources/pypy/site-packages/talon/voice.py", line 251, in ui_event
    self.update_scope()
  File "/Applications/Talon.app/Contents/Resources/pypy/site-packages/talon/voice.py", line 259, in update_scope
    self.reload()
  File "/Applications/Talon.app/Contents/Resources/pypy/site-packages/talon/voice.py", line 243, in reload
    engine.load(self.grammar)
  File "/Applications/Talon.app/Contents/Resources/pypy/site-packages/talon/engine.py", line 91, in load
    resp = grammar.load(self)
  File "/Applications/Talon.app/Contents/Resources/pypy/site-packages/talon/grammar.py", line 88, in load
    resp = self.cmd('g.load', main_rule=f'{self.name}::main', data=self.grammar.data)
  File "/Applications/Talon.app/Contents/Resources/pypy/site-packages/talon/grammar.py", line 113, in cmd
    return self.engine.cmd(cmd, **kwargs)```

**Comment by tuomassalo** [comment]
As of 0.0.6.0, Dragon dies so that I get the beach ball icon when hovering over its UI items.

**Comment by tuomassalo** [comment]
Still with us in 0.0.6.1.

**Comment by eliteramen** [comment]
@lunixbochs has identified that I am experiencing this issue as well.

@tuomassalo, does Dragon itself stop picking up your microphone? I would assume so because this issue has been categorized as a "Dragon seize" on Slack. Curious as to how often you encounter the problem you as well.

**Comment by tuomassalo** [comment]
@eliteramen - Dragon stops responding to e.g. 'wake up'. I'm facing the problem 2 or 3 times a day.

**Comment by eliteramen** [comment]
@tuomassalo I wonder what our environments have in common. It sounds like not many others are encountering this, or at least they are not talking about it.

Did you ever clean up VoiceCode changes? I deleted the Dragon plist as per @lunixbochs's recommendation, but I have yet to clean out the rest. Since you wake up Dragon, sounds like you use it or VC directly still in conjunction with Talon? (since Talon works in sleep mode)

**Comment by tuomassalo** [comment]
@eliteramen I've never had VoiceCode installed on this machine.

I have two profiles in Dragon. (One mic home and one at work.)

I've chosen British English in Dragon.

I can't think of any other special in my environment. I've edited the Dragon vocabulary very little, added and deleted just a few words.

**Comment by eliteramen** [comment]
I have american english in dragon.

I have several profiles in dragon, two of which i use regularly, and the others not at all (leftover from vc).

i have not done any edits to the dragon vocabulary directly. I am using a bunch of commands built on top of the examples as well as TerjeB's code.

@lunixbochs believes this is related to context switching; if so, changing all commands to global may mitigate this. Perhaps worth trying so that we can get some additional data here.

**Comment by eliteramen** [comment]
Workaround for now, restart script for hangs:

```
pkill -9 Dragon Talon; sleep 0.25; open -a Talon; open -a Dragon
```
Alternatively, you can attempt to mitigate the problem:
```
aegis [1:21 PM]
reduce contexts or context switches and wait for a fix.
```
In other words,
1. have fewer contexts for your commands (avoid `Context('git', ...)` AND `Context('std', ...)` AND `Context('chrome', ...)`)
    - clarification: if *none* of them have conditionals (e.g., the second argument to `Context('git', func=lambda app, win: any(t in app.bundle for t in terminals))`, then they are technically in the same "global" context
2. try not to switch applications too often

**Comment by eliteramen** [comment]
#19 may be related, according to @lunixbochs

**Comment by lunixbochs** [comment]
This was fixed in 0.0.6.12
